mitxgraders/helpers/mathfunc.py

```diff
@@ -66,26 +64,27 @@ def scalar_domain(display_name):
     Traceback (most recent call last):
     ValueError: Decorator 'scalar_domain' can only be used with unary functions.
 
-    Comment: The n-argument case seems intractable because it breaks get_number_of_args,
-    which is used in calc.py. But we could hard-code the n=1, n=2, n=3 cases.
-    Anything larger probably wouldn't come up in practice.
+    Comment: The n-argument case seems intractable because the _decorated_func
+    will be passed to get_number_of_args^[1], and so cannot have variadic signature *args.
+    But we could hard-code the n=1, n=2, n=3 cases.
+    [1]: In calc.py, eval_func calls is_callable_with, which uses get_number_of_args
```

Comment:
Ah, ok, I see what's going on now. The decorator would work fine with *args, but get_number_of_args would later be called on the decorator, see only *args, and say "Zero!". Proposed solution: add a `nin` attribute to the decorator to specify the number of arguments that are expected.

Reply:
So I thought about this a bit more. 


In calc.py's `eval_function` action:
1. Currently, we use `is_callable_with_args` to check that every function will be called with the correct number of arguments. This has been a pain in the but because `is_callable_with_args` usings `get_number_of_args`, which is hard to make general.
2. But if we have decorated some functions with `@specify_domain`, then the decorated function knows how many arguments it should have and can validate that itself. This function would not need to be validated by `is_callable_with_args`. 

With that in mind, I implemented a version of `@specify_domain` using voluptuous. Here's the interface:

```python
@specify_domain(1, [2, 2], 2, display_name='puppy')
def f(x, y, z):
    return x*y*z

print(f.validated)
# True       specify_domain adds a .validated property to f, so that
#            eval_function knows that f does its own validation.

a = MathArray([1, 2])
b = MathArray([[1, 2, 3], [4, 5, 6]])
c = MathArray([1, 2])
f(a, b, c)
# DomainError: There was an error evaluating function puppy(...)
# Input 0 received a vector of length 2, expected a scalar
# Input 1 received a matrix of shape (rows: 2, cols: 3), expected matrix of shape (rows: 2, cols: 2)
# Input 2 is ok.
```

Thoughts? It would probably take me a few hours to write documentation and tests, so I wanted your thoughts first.

Reply:
I like the sound of this - it's much better than the hacky way we're using to validate functions, and we can expose the decorator too (or fall back to our hacky methods) for user functions.